sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1355,7 +1356,8 @@ object CombineUnions extends Rule[LogicalPlan] {
     while (stack.nonEmpty) {
       stack.pop() match {
         case p1 @ Project(_, p2: Project)
-            if canCollapseExpressions(p1.projectList, p2.projectList, alwaysInline = false) =>
+            if canCollapseExpressions(p1.projectList, p2.projectList, alwaysInline = false) &&
+              !p1.projectList.exists(hasSubquery) && !p2.projectList.exists(hasSubquery) =>
```

Comment:
`hasSubquery` -> `hasCorrelatedSubquery`. I think non-correlated subqueries should work but let's add a test case just to be sure.

Reply:
+1